draft-ietf-resumable-uploads-protocol.md

```diff
@@ -109,126 +114,237 @@ The uploading of a file using the Resumable Uploads Protocol consists of multipl
+
+```
++---------+                                                     +---------+
+| Client  |                                                     | Server  |
++---------+                                                     +---------+
+     |                                                               |
+     | PATCH with Upload-Token, Upload-Offset, and Upload-Incomplete |
+     |-------------------------------------------------------------->|
+     |                                                               |
+     |              201 Created with Upload-Incomplete on completion |
+     |<--------------------------------------------------------------|
+     |                                                               |
+```
+
+2) The last chunk of the upload does not have the `Upload-Incomplete` header.
+
+```
++---------+                                 +---------+
+| Client  |                                 | Server  |
++---------+                                 +---------+
+     |                                           |
+     | PATCH with Upload-Token and Upload-Offset |
+     |------------------------------------------>|
+     |                                           |
+     |                 201 Created on completion |
+     |<------------------------------------------|
+     |                                           |
+```
+
 
 ## Upload Creation Procedure
 
-In order to initiate a resumable upload, the client MUST send a `POST` request to a known upload URL, under which endpoint the server supports resumable uploads as laid out in this document. Possible approaches for the client to obtain this upload URL as presented in the section Service Discovery. This `POST` request MUST include the `Client-Upload-Token` header and have it set to a unique value (TODO: More constraint to ensure its uniqueness).
+Upload Creation Procedure is designed to be compatible with a regular upload, with the intention that the client MAY initiate a resumable upload without the knowledge of server support. Therefore all methods which allow the request body are allowed (including `PATCH`), along with all response status codes. This procedure is identified with the presence of the `Upload-Token` header and the absence of the `Upload-Offset` header [Request Identification]. The client is RECOMMENDED to use `POST` request if not otherwise specified.
 
-The client SHOULD include the file's content in the request body and the server MUST attempt to store as much data as possible to minimize the data loss if the connection between the client and server get interrupted.
+The request MUST include the `Upload-Token` header which is a binary token with a minimum of 256-bit (16 byte) cryptographically-secure random binary data. The request MUST NOT include the `Upload-Offset` header or the `Upload-Incomplete` header. The server SHOULD reject shorter tokens by sending a `400 (Bad Request)` response.
```

Comment:
Specifying the cryptographic security seems a bit beyond scope here. Could that be a recommendation rather than part of the larger MUST statement? For example, one may want to make this a system-specific unique ID via some other means and cryptographically sign it or secure it in some other way, rather than being required to include a random binary token.